test/support/lib/thrift_test_case.ex

```diff
@@ -1,131 +1,134 @@
+      Module.register_attribute(__MODULE__, :thrift_record_modules, accumulate: true)
     end
   end
 
-  defmacro __before_compile__(env) do
-    tag = Module.get_attribute(__CALLER__.module, :moduletag)
+  def implement?(module) do
+    tag = Module.get_attribute(module, :moduletag)
     |> Map.new(fn tag ->  {tag, true} end)
 
     config = ExUnit.configuration
     case ExUnit.Filters.eval(config[:include], config[:exclude], tag, []) do
       :ok ->
-        compile_and_build_erlang_helpers(__CALLER__, env)
+        true
       {:error, _} ->
-        nil
+        false
     end
   end
 
-  defp compile_and_build_erlang_helpers(caller, env) do
-    opts = Module.get_attribute(caller.module, :thrift_test_opts)
+  def quoted_contents(module, contents) do
+    compile_and_build_helpers(module)
+    directives = quoted_directives(module)
+    [directives, contents]
+  end
 
-    namespace = inspect(env.module)
+  defp compile_and_build_helpers(module) do
+    files = get_thrift_files(module)
+    dir = Module.get_attribute(module, :thrift_test_dir)
+    opts = Module.get_attribute(module, :thrift_test_opts)
+    generate_files(files, module, dir, opts)
+  end
 
-    out_root = Path.join(@project_root, "tmp")
-    dir = Path.join([out_root, inspect(__MODULE__), namespace])
-    File.rm_rf!(dir)
-    File.mkdir_p!(dir)
+  defp get_thrift_files(module) do
+    files = Module.get_attribute(module, :thrift_file)
+    Module.delete_attribute(module, :thrift_file)
+    Module.register_attribute(module, :thrift_file, [accumulate: true])
 
+    Enum.reverse(files)
+  end
 
-    modules = caller.module
-    |> Module.get_attribute(:thrift_file)
-    |> Enum.reverse
-    |> Enum.map(fn [name: filename, contents: contents] ->
-      filename = Path.expand(filename, dir)
-      File.write!(filename, "namespace elixir #{namespace}\n" <> contents)
-      filename
-    end)
-    |> Enum.flat_map(&Thrift.Generator.generate!(&1, dir))
-    |> Enum.uniq
-    |> Enum.map(fn output_file ->
-      output_file
+  defp write_thrift_file(config, namespace, dir) do
+    filename = config
+      |> Keyword.fetch!(:name)
       |> Path.expand(dir)
-      |> Code.eval_file
 
-      namespace_module = output_file
-      |> Path.dirname
-      |> String.split("/")
-      |> Enum.map(&Macro.camelize/1)
-      |> Enum.join(".")
+    contents = Keyword.fetch!(config, :contents)
 
-      basename_module = output_file
-      |> Path.basename(".ex")
-      |> Macro.camelize
+    File.write!(filename, "namespace elixir #{inspect namespace}\n" <> contents)
 
-      :"Elixir.#{namespace_module}.#{basename_module}"
-    end)
+    filename
+  end
+
+  defp require_file(file, dir) do
+    case Code.require_file(file, dir) do
+      nil ->
+        []
+      modules ->
+        modules = Enum.map(modules, fn({module, _}) -> Module.split(module) end)
+        Enum.filter_map(modules, &alias?(&1, modules), &Module.concat/1)
```

Comment:
That isn't quite equivalent, but it just shows how confusing that piece of code is. We need the rebound `modules` as second argument to `alias?/2`.